nft_ingester/src/processors/transaction_processor.rs

```diff
@@ -3,68 +3,69 @@ use std::{sync::Arc, time::Duration};
 use chrono::Utc;
 use interface::signature_persistence::UnprocessedTransactionsGetter;
 use metrics_utils::MessageProcessMetricsConfig;
-use tokio::{
-    sync::{broadcast::Receiver, Mutex},
-    task::{JoinError, JoinSet},
-    time::sleep as tokio_sleep,
-};
+use tokio_util::sync::CancellationToken;
 use tracing::error;
 
 use super::transaction_based::bubblegum_updates_processor::BubblegumTxProcessor;
 use crate::{error::IngesterError, redis_receiver::get_timestamp_from_id};
 
 const TRANSACTIONS_GETTER_IDLE_TIMEOUT_MILLIS: u64 = 250;
 
-pub async fn run_transaction_processor<TG>(
-    rx: Receiver<()>,
-    mutexed_tasks: Arc<Mutex<JoinSet<Result<(), JoinError>>>>,
+pub fn run_transaction_processor<TG>(
+    cancellation_token: CancellationToken,
     unprocessed_transactions_getter: Arc<TG>,
     geyser_bubblegum_updates_processor: Arc<BubblegumTxProcessor>,
     message_process_metrics: Option<Arc<MessageProcessMetricsConfig>>,
 ) where
     TG: UnprocessedTransactionsGetter + Send + Sync + 'static,
 {
-    let run_transaction_processor = async move {
-        while rx.is_empty() {
-            let txs = match unprocessed_transactions_getter.next_transactions().await {
-                Ok(txs) => txs,
-                Err(err) => {
-                    error!("Get next transactions: {}", err);
-                    tokio_sleep(Duration::from_millis(TRANSACTIONS_GETTER_IDLE_TIMEOUT_MILLIS))
-                        .await;
-                    continue;
-                },
-            };
+    usecase::executor::spawn(async move {
+        cancellation_token
+            .run_until_cancelled(async move {
+                loop {
+                    let txs = match unprocessed_transactions_getter.next_transactions().await {
+                        Ok(txs) => txs,
+                        Err(err) => {
+                            error!("Get next transactions: {}", err);
+                            tokio::time::sleep(Duration::from_millis(
+                                TRANSACTIONS_GETTER_IDLE_TIMEOUT_MILLIS,
+                            ))
+                            .await;
+                            continue;
+                        },
+                    };
 
-            for tx in txs {
-                match geyser_bubblegum_updates_processor.process_transaction(tx.tx, false).await {
-                    Ok(_) => {
-                        if let Some(message_process_metrics) = &message_process_metrics {
-                            if let Some(message_timestamp) = get_timestamp_from_id(&tx.id) {
-                                let current_timestamp = Utc::now().timestamp_millis() as u64;
+                    for tx in txs {
+                        match geyser_bubblegum_updates_processor
+                            .process_transaction(tx.tx, false)
```

Comment:
run_until_cancelled assumes the future is safe to cancel, but if you're canceled in the middle of process_transaction - we'll have an issue here